pilot/pkg/bootstrap/server.go

```diff
@@ -426,43 +432,90 @@ func (c *mockController) AppendInstanceHandler(f func(*model.ServiceInstance, mo
-func (s *Server) initConfigController(args *PilotArgs) error {
-	mcpServerAddrs := args.MCPServerAddrs
+func (s *Server) initMCPConfigController(args *PilotArgs) error {
+	clientNodeID := ""
+	supportedTypes := make([]string, len(model.IstioConfigTypes))
+	for i, model := range model.IstioConfigTypes {
+		supportedTypes[i] = fmt.Sprintf("type.googleapis.com/%s", model.MessageName)
+	}
 
-	if len(mcpServerAddrs) > 0 {
-		clientNodeID := args.Namespace
-		supportedTypes := make([]string, len(model.IstioConfigTypes))
-		for i, model := range model.IstioConfigTypes {
-			supportedTypes[i] = fmt.Sprintf("type.googleapis.com/%s", model.MessageName)
-		}
+	options := coredatamodel.Options{
+		DomainSuffix: args.Config.ControllerOptions.DomainSuffix,
+	}
+	mcpController := coredatamodel.NewController(options)
 
-		mcpController := coredatamodel.NewController()
+	ctx, cancel := context.WithCancel(context.Background())
+	var clients []*mcpclient.Client
 
-		s.addStartFunc(func(stop <-chan struct{}) error {
-			ctx, cancel := context.WithCancel(context.Background())
-			go func() {
-				<-stop
-				cancel()
-			}()
-
-			for _, addr := range mcpServerAddrs {
-				// TODO: make this a secure connection before shipping
-				conn, err := grpc.Dial(addr, grpc.WithInsecure())
-				if err != nil {
-					log.Errorf("Unable connecting to MCP Server: %v\n", err)
-					return err
+	for _, addr := range args.MCPServerAddrs {
+		u, err := url.Parse(addr)
+		if err != nil {
+			return err
+		}
+
+		securityOption := grpc.WithInsecure()
+		if u.Scheme == "mcp" {
+			requiredFiles := []string{
+				args.MCPCredentialOptions.CertificateFile,
+				args.MCPCredentialOptions.KeyFile,
+				args.MCPCredentialOptions.CACertificateFile,
+			}
+			log.Infof("Secure MCP configured. Waiting for required certificate files to become available: %v",
+				requiredFiles)
+			for len(requiredFiles) > 0 {
+				if _, err := os.Stat(requiredFiles[0]); os.IsNotExist(err) {
+					log.Infof("%v not found. Checking again in %v", requiredFiles[0], requiredMCPCertCheckFreq)
+					select {
+					case <-ctx.Done():
+						return ctx.Err()
+					case <-time.After(requiredMCPCertCheckFreq):
+						// retry
+					}
+					continue
 				}
+				log.Infof("%v found", requiredFiles[0])
+				requiredFiles = requiredFiles[1:]
+			}
 
-				cl := mcpapi.NewAggregatedMeshConfigServiceClient(conn)
-				mcpClient := mcpclient.New(cl, supportedTypes, mcpController, clientNodeID, map[string]string{})
-				go mcpClient.Run(ctx)
+			watcher, err := creds.WatchFiles(ctx.Done(), args.MCPCredentialOptions)
+			if err != nil {
+				return err
 			}
-			return nil
-		})
+			credentials := creds.CreateForClient(u.Hostname(), watcher)
+			securityOption = grpc.WithTransportCredentials(credentials)
+		}
+		conn, err := grpc.DialContext(ctx, u.Host, securityOption)
+		if err != nil {
+			log.Errorf("Unable connecting to MCP Server: %v\n", err)
+			return err
```

Comment:
continue

If one server is down for some reason, we should still connect to the other servers.

You probably also need to extract a function for connecting to one server, with some retry logic.

Reply:
The grpc client library handles (re)dialing the server on connection failures. The MCP client handles [re-establishing the MCP client stream](https://github.com/istio/istio/blob/cdfc167409363825668a40e70b4d2ae3ea903dbe/pkg/mcp/client/client.go#L256-L272) when the underlying connection is lost.

I believe non-blcoking Dial should only fail if the client is mis-configured (e.g. secure with missing transport credentials).